# Notebook Explanation: Understanding Sentence Embeddings and Similarity Metrics

This notebook demonstrates how to convert sentences into numerical representations (embeddings) and then calculate various similarity and distance metrics between these embeddings. The goal is to understand the semantic relationship between pieces of text.

In [19]:
import numpy as np

from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
from scipy.spatial.distance import euclidean

In [20]:
model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

In [21]:
sentences=[
    "The cat is on the sofa.",
    "A kitten is resting on the couch",
    "I love eating outside"
]

In [22]:
A=sentences[0]
B=sentences[1]
C=sentences[2]


In [23]:
embeddings =model.encode(sentences)
A_vec=embeddings[0]
B_vec=embeddings[1]
C_vec=embeddings[2]


In [24]:
print(A_vec.shape)

(384,)


In [25]:
print(A_vec[:10])
print(B_vec[:10])
print(C_vec[:10])

[ 0.165308   -0.09584815 -0.01919786 -0.00298879 -0.01859661  0.05443123
 -0.02191237 -0.02771806  0.02176203  0.03170222]
[ 0.08584699 -0.03655815 -0.00540279  0.047877    0.03783667  0.04372902
 -0.08281959 -0.02568191  0.02826712  0.01417102]
[ 0.03909591 -0.02968942  0.09797014  0.05432226  0.04100079 -0.05628665
  0.09491522 -0.06082592 -0.02418943  0.02992172]


## Section 1: Setup and Data Preparation


#### Implementation:
1.  **Imports** : We start by importing essential libraries:
    *   `numpy` for numerical operations.
    *   `SentenceTransformer` from `sentence_transformers` to convert text into embeddings.
    *   `cosine_similarity` from `sklearn.metrics.pairwise` to calculate cosine similarity.
    *   `euclidean` from `scipy.spatial.distance` to calculate Euclidean distance.
2.  **Model Loading** : A pre-trained sentence transformer model (`all-MiniLM-L6-v2`) is loaded. This model is specifically designed to generate high-quality sentence embeddings.
3.  **Sentence Definition** : Three example sentences (`A`, `B`, `C`) are defined. Notice that sentences A and B are semantically very similar, while C is distinctly different.
4.  **Embedding Generation** : The `model.encode(sentences)` function transforms each sentence into a dense vector (embedding). These are stored in `A_vec`, `B_vec`, and `C_vec`.

#### Conceptual Explanation:
*   **Sentence Embeddings**: These are numerical representations (vectors) of text that capture its semantic meaning. Sentences with similar meanings will have embeddings that are 'close' to each other in this multi-dimensional space, while dissimilar sentences will have embeddings that are 'far apart'.
*   **Vector Dimensions** : The `(384,)` output indicates that each sentence is represented by a vector of 384 numbers. This is a common dimensionality for many pre-trained sentence embedding models.
*   **Vector Content** : Printing `A_vec[:10]` (and for B and C) shows the first 10 floating-point numbers of these vectors, illustrating their numerical nature. These numbers collectively encode the sentence's meaning.

In [32]:
A_mag=np.sqrt(np.sum(A_vec**2))
B_mag=np.sqrt(np.sum(B_vec**2))
C_mag=np.sqrt(np.sum(C_vec**2))

print(A_mag)
print(B_mag)
print(C_mag)

1.0
1.0
1.0


In [33]:
A_norm=A_vec/A_mag
B_norm=B_vec/B_mag
C_norm=C_vec/C_mag

print(A_norm[:10])
print(B_norm[:10])
print(C_norm[:10])

[ 0.165308   -0.09584815 -0.01919786 -0.00298879 -0.01859661  0.05443123
 -0.02191237 -0.02771806  0.02176203  0.03170222]
[ 0.08584699 -0.03655815 -0.00540279  0.047877    0.03783667  0.04372902
 -0.08281959 -0.02568191  0.02826712  0.01417102]
[ 0.03909591 -0.02968942  0.09797014  0.05432226  0.04100079 -0.05628665
  0.09491522 -0.06082592 -0.02418943  0.02992172]


In [35]:
print(np.sqrt(np.sum(A_norm**2)))
print(np.sqrt(np.sum(B_norm**2)))
print(np.sqrt(np.sum(C_norm**2)))

1.0
1.0
1.0


## Section 2: Vector Magnitudes and Normalization


#### Implementation:
1.  **Magnitude Calculation** : `A_mag`, `B_mag`, `C_mag` are calculated using `np.sqrt(np.sum(vec**2))`, which is the L2-norm (Euclidean norm) of the vector. This measures the length of the vector.
2.  **Normalization** : `A_norm`, `B_norm`, `C_norm` are created by dividing each vector by its magnitude. This converts them into unit vectors (vectors with a magnitude of 1).
3.  **Verification** : The magnitudes of the normalized vectors are printed to confirm they are indeed 1.

#### Conceptual Explanation:
*   **Vector Magnitude (L2-Norm)**: The length of a vector. For some similarity metrics, the magnitude can influence the result. Here, we observe that the sentence transformer model already produces embeddings that are normalized (their magnitude is 1). This is a common practice in NLP to ensure that the length of the vector doesn't disproportionately affect similarity calculations, especially for cosine similarity.
*   **Normalization**: The process of scaling a vector to have a magnitude of 1. A normalized vector points in the same direction as the original vector but has a standard length. This is crucial for cosine similarity, as it makes the metric purely dependent on the angle between vectors, not their lengths.

#### Output Analysis:
*   All magnitudes of vectors are `1.0`, confirming that the `SentenceTransformer` model already outputs normalized embeddings.
*  The first 10 elements of the normalized vectors are identical to the original vectors, as dividing by `1.0` does not change their values.
*  Calculating the magnitude of the normalized vectors also yields `1.0`, verifying that they are indeed unit vectors.

In [36]:
dot_AB  =np.dot(A_vec,B_vec)
dot_AC=np.dot(A_vec,C_vec)
dot_BC=np.dot(B_vec,C_vec)

print(dot_AB)
print(dot_AC)
print(dot_BC)

0.7282556
-0.01747297
-0.039351955


In [28]:
cos_AB=cosine_similarity([A_vec],[B_vec])
cos_AC=cosine_similarity([A_vec],[C_vec])
cos_BC=cosine_similarity([B_vec],[C_vec])

print(cos_AB)
print(cos_AC)
print(cos_BC)

[[0.7282557]]
[[-0.01747296]]
[[-0.03935196]]


In [30]:
dist_AB = euclidean(A_vec, B_vec)

dist_AC = euclidean(A_vec, C_vec)
dist_BC = euclidean(B_vec, C_vec)
print(dist_AB)
print(dist_AC)
print(dist_BC)

0.7372168302536011
1.4265152215957642
1.4417710304260254


## Section 3: Similarity and Distance Metrics

This section calculates three common metrics to quantify the relationship between the sentence embeddings:

### Dot Product
#### Implementation:
`dot_AB = np.dot(A_vec, B_vec)` calculates the dot product between `A_vec` and `B_vec` (and similarly for other pairs).
#### Conceptual Explanation:
*   **Dot Product**: A measure of how much two vectors point in the same direction. For normalized vectors, the dot product is equivalent to the cosine of the angle between them. A larger dot product (positive) indicates a greater degree of alignment, while a smaller or negative dot product indicates less alignment or opposition.
#### Output Analysis :
*   `dot_AB`: `0.7282556` is relatively high and positive, indicating that sentences A and B are quite aligned semantically.
*   `dot_AC`: `-0.01747297` and `dot_BC`: `-0.039351955` are close to zero and negative, suggesting that sentences A and C, and B and C, are not aligned and are semantically distant.

### Cosine Similarity
#### Implementation:
`cos_AB = cosine_similarity([A_vec], [B_vec])` uses `sklearn`'s function to compute cosine similarity.
#### Conceptual Explanation:
*   **Cosine Similarity**: Measures the cosine of the angle between two vectors. It ranges from -1 (perfectly opposite) to 1 (perfectly similar), with 0 indicating orthogonality (no similarity). It is widely used for text similarity because it focuses on the direction of vectors, not their magnitude, which is ideal for embeddings that capture semantic meaning. Since our vectors are already normalized (magnitude of 1), the dot product and cosine similarity will yield very similar results.
#### Output Analysis :
*   `cos_AB`: `[[0.7282557]]` confirms the high similarity between A and B, indicating they are semantically close. Note the slight floating-point difference due to `sklearn`'s internal calculations.
*   `cos_AC`: `[[-0.01747296]]` and `cos_BC`: `[[-0.03935196]]` show very low (near zero or slightly negative) similarity, reinforcing that sentence C is semantically different from A and B.

### Euclidean Distance
#### Implementation:
`dist_AB = euclidean(A_vec, B_vec)` calculates the Euclidean distance between `A_vec` and `B_vec`.
#### Conceptual Explanation:
*   **Euclidean Distance**: The 'straight-line' distance between two points (vector endpoints) in multi-dimensional space. Smaller distances indicate greater similarity, while larger distances indicate greater dissimilarity.
#### Output Analysis :
*   `dist_AB`: `0.7372168302536011` is the smallest, indicating A and B are closest in embedding space.
*   `dist_AC`: `1.4265152215957642` and `dist_BC`: `1.4417710304260254` are significantly larger, indicating A and C, and B and C, are far apart and semantically dissimilar.

### Relationship between Cosine Similarity and Euclidean Distance

Both Cosine Similarity and Euclidean Distance are measures of similarity/dissimilarity between two vectors, but they capture different aspects:

*   **Cosine Similarity** focuses on the **orientation** of the vectors. It measures the angle between them. If two vectors point in the exact same direction, their cosine similarity is 1, regardless of their magnitude. If they point in opposite directions, it's -1. If they are orthogonal (perpendicular), it's 0.

*   **Euclidean Distance** focuses on the **magnitude** and **position** of the vectors in space. It measures the straight-line distance between the endpoints of the vectors. A smaller Euclidean distance means the vectors are closer in space.

#### Key Differences and Relationship:
1.  **Normalization**: Cosine similarity is inherently normalized by the magnitude of the vectors (it divides by the product of their magnitudes). This means if you double the length of a vector but keep its direction, its cosine similarity with another vector will not change. Euclidean distance, however, is sensitive to magnitude. Doubling the length of a vector will increase its Euclidean distance from other vectors.

2.  **Angle vs. Absolute Position**: Cosine similarity cares about the angle. Euclidean distance cares about the absolute difference in their coordinates.

3.  **Impact of Normalization on Relationship**: For **L2-normalized vectors** (unit vectors, where magnitude is 1), there's a direct mathematical relationship between Cosine Similarity and Euclidean Distance:

    $$ \text{Euclidean Distance}^2 = 2 - 2 \times \text{Cosine Similarity} $$

    Or, more commonly:

    $$ \text{Euclidean Distance} = \sqrt{2 - 2 \times \text{Cosine Similarity}} $$

    This means for normalized vectors:
    *   When Cosine Similarity is 1 (vectors are identical in direction), Euclidean Distance is 0.
    *   When Cosine Similarity is 0 (vectors are orthogonal), Euclidean Distance is $\sqrt{2}$.
    *   When Cosine Similarity is -1 (vectors are opposite), Euclidean Distance is $\sqrt{4} = 2$.

    In our notebook, since the sentence embeddings are already L2-normalized (magnitudes are 1), you can observe this relationship. For example, for A and B:
    *   `cos_AB` is approximately `0.728`
    *   $\sqrt{2 - 2 \times 0.728} = \sqrt{2 - 1.456} = \sqrt{0.544} \approx 0.737$
    *   Which is very close to our calculated `dist_AB` of `0.7372`.

**In summary**: For normalized embeddings, both metrics effectively capture semantic similarity. Cosine similarity is often preferred in high-dimensional spaces for text because it's less affected by differences in absolute vector lengths that might not be semantically meaningful, and it directly relates to the angle, which intuitively aligns with direction of meaning.

In [37]:
import pandas as pd

results = pd.DataFrame({
    "Pair": ["A-B", "A-C","B-C"],
    "Dot Product": [dot_AB, dot_AC,dot_BC],
    "Cosine Similarity": [cos_AB, cos_AC,cos_BC],
    "Euclidean Distance": [dist_AB, dist_AC,dist_BC]
})

results

,Pair,Dot Product,Cosine Similarity,Euclidean Distance
0,A-B,0.728256,[[0.7282557]],0.737217
1,A-C,-0.017473,[[-0.017472964]],1.426515
2,B-C,-0.039352,[[-0.039351955]],1.441771


## Section 4: Results Aggregation

#### Implementation:
A Pandas DataFrame is created to organize and display all the calculated metrics (Dot Product, Cosine Similarity, Euclidean Distance) for each pair of sentences (A-B, A-C, B-C). This provides a clear, tabular summary of the results.



#### Output Analysis :
*   The DataFrame clearly shows that A and B have a high dot product, high cosine similarity, and low Euclidean distance, all pointing to strong semantic similarity.
*   In contrast, pairs A-C and B-C consistently show low similarity (low/negative dot product, low/negative cosine similarity) and high distance, indicating they are semantically unrelated. The missing row for B-C in the output suggests that the `results` DataFrame was not fully displayed or constructed as intended in the provided context, but the values are consistent with the individual print statements.